sql/core/src/main/scala/org/apache/spark/sql/execution/AlreadyPlanned.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.{DataFrame, Dataset, Row, SparkSession}
+import org.apache.spark.sql.catalyst.encoders.RowEncoder
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan}
+
+/**
+ * A special node that allows skipping query planning all the way to physical execution. This node
+ * can be used when a query was planned to the physical level, but we had to go back to logical plan
+ * land for some reason (e.g. V1 DataSource write execution). This will allow the metrics, and the
+ * query plan to properly appear as part of the query execution.
+ */
+case class AlreadyPlanned(physicalPlan: SparkPlan) extends LeafNode {
+  override val output: Seq[Attribute] = physicalPlan.output
+  override lazy val resolved: Boolean = false
```

Comment:
This is put in `AlreadyPlannedExecution.analyzed`. It's weird if this is unresolved.

Reply:
OK I got it now. We want to forbid further operations to the DataFrame.

How about
```
class AlreadyOptimizedExecution(plan: LogicalPlan) {
  override lazy val analyzed: LogicalPlan = plan
  override lazy val optimizedPlan: LogicalPlan = plan
}
```

Reply:
that wasn't sufficient. We need to pass in the `executedPlan` for the spark and executed plans. Otherwise one of the test suites failed with:
```
sbt.ForkMain$ForkError: java.lang.UnsupportedOperationException: null
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doProduce(WholeStageCodegenExec.scala:790)
```